lib/improver/lapse_rate.py

```diff
@@ -0,0 +1,308 @@
+
+        def calc_lapse_rate(temperature, orography, ind_central_point):
+            """
+            Least-squares fit to local temperature and altitude data to
+            calculate a local lapse rate.
+
+            """
+
+            # If central point NaN then return blank value.
+            if np.isnan(temperature[ind_central_point]):
+                return 0.0
+
+            # Remove NaN values before calculation.
+            y_data = temperature[~np.isnan(temperature)]
+            x_data = orography[~np.isnan(orography)]
+
+            # If empty array - return blank value.
+            if len(y_data) == 0:
+                return 0.0
+
+            matrix = np.stack([x_data, np.ones(len(x_data))], axis=0).T
+            gradient, intercept = lstsq(matrix, y_data)[0]
+
+            return gradient
+
+        def create_height_diff_mask(all_orog_subsections, ind_central_point,
+                                    max_height_diff):
+            """
+            Create a mask for any neighbouring points where the height
+            difference from the central point is greater than max_height_diff.
+
+            Slice through the orography subsection array to remove central
+            points. Then subtract central height value from neighbourhoods.
+            """
+            central_points = all_orog_subsections[:, ind_central_point]
+            central_points = np.swapaxes([central_points], 0, 1)
+
+            height_diff = np.subtract(all_orog_subsections, central_points)
+            height_diff = np.absolute(height_diff)
+
+            height_diff_mask = np.where(height_diff >= max_height_diff, True,
+                                        False)
+            return height_diff_mask
+
+        DALR = 0.0098  # Dry Adiabatic Lapse Rate in unit of K/m-1
+
+        # footprint=3 corresponds to a 3x3 array centred on central point.
+        # Exclude even numbers since the central point is lower right eg.
+        # 4x4 array means centre is (2,2).
+        if footprint_len % 2 == 0:
+            msg = "Footprint length is {}, must be odd number!"
+            raise ValueError(msg.format(footprint_len))
+
+        # generic_filter extracts the neighbourhood and returns a 1D array.
+        # ind_central_point indicates where the central point is on this array
+        footprintarray_size = footprint_len**2
+        ind_central_point = int(footprintarray_size/2)
+
+        if not isinstance(temperature_cube, iris.cube.Cube):
+            msg = "Temperature input is not a cube, but {}"
+            raise TypeError(msg.format(type(temperature_cube)))
+
+        if not isinstance(orography_cube, iris.cube.Cube):
+            msg = "Orography input is not a cube, but {}"
+            raise TypeError(msg.format(type(orography_cube)))
+
+        if not isinstance(land_sea_mask_cube, iris.cube.Cube):
+            msg = "Land/Sea mask input is not a cube, but {}"
+            raise TypeError(msg.format(type(land_sea_mask_cube)))
+
```

Comment:
There are no checks for diagnostic units here. We would not know if we are working in Fahrenheit and feet. All of these units can be converted, so rather than an exception, we might just want to work in the input units and perform a conversion at the end into K m-1.